airflow/providers/amazon/aws/fs/s3.py

```diff
@@ -85,7 +85,15 @@ def get_fs(conn_id: str | None) -> AbstractFileSystem:
     if proxy_uri := s3_service_config.get(S3_PROXY_URI, None):
         config_kwargs["proxies"] = {"http": proxy_uri, "https": proxy_uri}
 
-    fs = S3FileSystem(session=session, config_kwargs=config_kwargs, endpoint_url=endpoint_url)
+    anon = False
+    if (
+        aws.conn_config.aws_access_key_id is None
+        and aws.conn_config.aws_secret_access_key is None
+        and aws.conn_config.aws_session_token is None
+    ):
+        anon = True
```

Comment:
Unfortunetly it can't be a solution for indicate is lack of credentials.
Credentials could be provided from assuming role by use EC2/ECS/EKS IRSA, environment variables, profiles combination of them and even by custom federation.

I think the only good way it is try to call `session.get_credentials()` if it is return None than no credentials provided.

Or create separate key into the `service_config.s3` which are use for determine is required anonymous access or not.

Reply:
I'm not sure, but it also could be a situation that Anonymous access already works, by provide `signature_version: "unsigned"` into the [AWS connection extra `config_kwargs`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html#configuring-the-connection).

AFAIK, anonymous access implemented in AWS API / `botocore` / `boto3` just by no signing requests to API

Reply:
I'm okay with replacing it with checking for `session.get_credentials()`. The issue is that with an empty connection id and no env variables it will not default to anonymous, which imho it should. So adding it to connection extra doesn't makes sense I think.

Reply:
> The issue is that with an empty connection id and no env variables it will not default to anonymous,

`None` = use [default `boto3` credentials strategy](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials), and this is not mean use anonymous. Depend on situation credentials check might work, and might not work.

E.g. user deploy on EC2 with enabled Instance IAM Profile, in this case credentials might be empty, but session.get_credentials() return something different than `None`

And even more funny stuff: for historical reason AWS Connection fallback to boto3 default strategy even if connection id not found which is personally for me "Have a nice debugging time" 🙄 

I'm not sure, but maybe better to check config_kwargs.get("signature_version") == "unsigned". 
However it is also doesn't guarantee that we correctly guess is it required or not. All possible cases might be only resolved when we get finally constructed `botocore.config.Config` in `AwsGenericHook. _get_config(config=None)`

https://github.com/apache/airflow/blob/1aa91a482630f389005a9a36b0f038e675163411/airflow/providers/amazon/aws/hooks/base_aws.py#L638-L654

Reply:
So sorry. it take some time to complete my additional comment. 😢

Reply:
Here we are trying to to do this

```python
x = ObjectStoragePath("s3://xxxx")
```

There is no way to pass `config_kwargs` to this at the moment, and I don't think that makes sense for this use case like I said.  

However, note that passing `None` here does use the default credentials strategy of boto3, which might return *empty* credentials. I am only setting `anon=True` if no credentials are returned. So I think we're good?

Reply:
> There is no way to pass config_kwargs to this at the moment

But you use AWS Connection thought the hook, so users definitely could provide `config_kwargs`

https://github.com/apache/airflow/blob/5fe65b7b1d205cf465e1ed2200bbe4d3893f145a/airflow/providers/amazon/aws/fs/s3.py#L58

I'm not sure the best way to resolve all cases, but seems like we need to have explicit settings for this case in connection and documented it. Because right now I can't see any word about `ObjectStoragePath("s3://xxxx") `into the Amazon documentation, how to configure it, and how to use it. Without docs it could be hard to triangle issues, answer on discussions in GH or Slack

Reply:
I agree with you that this can be better documented, but it should be done in the ObjectStorage documentation and not as part of a provider and as such I think it is outside the scope of this PR. 

I am happy to improve upon the documentation in a separate pr, but can also use your assistance there due to your knowledge of this area.